subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmFeature.java

```diff
@@ -89,133 +115,169 @@ public DefaultJvmFeature(
+        this.compileClasspath = configurations.getByName(sourceSet.getCompileClasspathConfigurationName());
 
-        // TODO: compileOnly should probably also extend from compileOnlyApi
-        this.implementation.extendsFrom(api);
+        PublishArtifact jarArtifact = new LazyPublishArtifact(jar, project.getFileResolver(), project.getTaskDependencyFactory());
+        this.apiElements = createApiElements(configurations, jarArtifact, compileJava, elementsConfigurationRole);
+        this.runtimeElements = createRuntimeElements(configurations, jarArtifact, compileJava, elementsConfigurationRole);
 
-        this.apiElements = createApiElements(configurations, apiElementsConfigurationName, jar, compileJava, elementsConfigurationRole);
-        this.runtimeElements = createRuntimeElements(configurations, runtimeElementsConfigurationName, jar, compileJava, elementsConfigurationRole);
-
-        // TODO: This behavior is weird. Specifically for the main source set, we do this thing where
-        // main classpaths "extend" the new feature. What is the use case for this? Any new features would
-        // get their own buckets, but share the same source set, compile task, jar task, sources jar, and javadoc jar.
-        // The documentation provides an example of creating multiple features with the main source set, but the
-        // use case is still confusing to consider. It seems this is some overly complex alternative to optional dependencies. See:
-        // https://docs.gradle.org/current/userguide/feature_variants.html#sec:feature_variant_source_set
-        if (SourceSet.isMain(sourceSet)) {
-            // we need to wire the compile only and runtime only to the classpath configurations
-            configurations.getByName(sourceSet.getCompileClasspathConfigurationName()).extendsFrom(implementation, compileOnly);
-            configurations.getByName(sourceSet.getRuntimeClasspathConfigurationName()).extendsFrom(implementation, runtimeOnly);
-            // and we also want the feature dependencies to be available on the test classpath
-            configurations.getByName(JvmConstants.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME).extendsFrom(implementation, compileOnlyApi);
-            configurations.getByName(JvmConstants.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME).extendsFrom(implementation, runtimeOnly);
+        if (extendProductionCode) {
+            doExtendProductionCode();
         }
 
         JavaPluginExtension javaPluginExtension = project.getExtensions().findByType(JavaPluginExtension.class);
-        JvmPluginsHelper.configureJavaDocTask(displayName, sourceSet, tasks, javaPluginExtension);
+        JvmPluginsHelper.configureJavaDocTask("'" + name + "' feature", sourceSet, tasks, javaPluginExtension);
+    }
+
+    /**
+     * This method is one of the primary reasons that we want to deprecate the "extending" behavior. It updates
```

Comment:
This is no doubt useful information, but is it really something that should be in a Javadoc?

Reply:
Where would you put it? Comments in the method body?

Reply:
Yeah, lacking a better idea. It should be really in a design document/spec somewhere.

Reply:
I will move this to the method body. Prior to moving forward with deprecation, I will make sure to we document this in a more formal design document in order to document the decision process.